pkg/rest/rate_limit.go

```diff
@@ -0,0 +1,131 @@
+
+		limit, humanReadableTimestamp, timeInMicroseconds, err := s.extractRateLimitHeaders(response)
+		if err != nil {
+			return response, err
+		}
+
+		util.Log.Info("Rate limit of %d requests/min reached: Applying rate limit strategy (simpleSleepRateLimitStrategy, iteration: %d)", limit, currentIteration+1)
+		util.Log.Info("simpleSleepRateLimitStrategy: Attempting to sleep until %s", humanReadableTimestamp)
+
+		// Attention: this uses client time:
+		now := timelineProvider.Now()
+
+		// Attention: this uses server time:
+		resetTime := s.convertMicrosecondsToUnixTime(timeInMicroseconds)
+
+		// Attention: this mixes client and server time:
+		sleepDuration := resetTime.Sub(now)
+		util.Log.Debug("simpleSleepRateLimitStrategy: Calculated sleep duration of %f seconds...", sleepDuration.Seconds())
+
+		// That's why we need plausible min/max wait time defaults:
+		sleepDuration = s.applyMinMaxDefaults(sleepDuration)
+
+		util.Log.Debug("simpleSleepRateLimitStrategy: Sleeping for %f seconds...", sleepDuration.Seconds())
+		timelineProvider.Sleep(sleepDuration)
+		util.Log.Debug("simpleSleepRateLimitStrategy: Slept for %f seconds", sleepDuration.Seconds())
+
+		// Checking again:
+		currentIteration++
+		response = callback()
+	}
+
+	return response, nil
+}
+
+func (s *simpleSleepRateLimitStrategy) extractRateLimitHeaders(response Response) (limit string, humanReadableResetTimestamp string, resetTimeInMicroseconds int64, err error) {
+
+	limitAsArray := response.Headers["X-RateLimit-Limit"]
+	resetAsArray := response.Headers["X-RateLimit-Reset"]
+
+	if limitAsArray == nil || limitAsArray[0] == "" {
+		return "", "", 0, errors.New("rate limit header 'X-RateLimit-Limit' not found")
+	}
+	if resetAsArray == nil || resetAsArray[0] == "" {
+		return "", "", 0, errors.New("rate limit header 'X-RateLimit-Reset' not found")
+	}
+
+	limit = limitAsArray[0]
+	humanReadableResetTimestamp, resetTimeInMicroseconds, err = util.StringTimestampToHumanReadableFormat(resetAsArray[0])
+	if err != nil {
+		return "", "", 0, err
+	}
+
+	return limit, humanReadableResetTimestamp, resetTimeInMicroseconds, nil
+}
+
+func (s *simpleSleepRateLimitStrategy) applyMinMaxDefaults(sleepDuration time.Duration) time.Duration {
+
+	minWaitTimeInNanoseconds := 5 * time.Second
+	maxWaitTimeInNanoseconds := 1 * time.Minute
+
+	if sleepDuration.Nanoseconds() < minWaitTimeInNanoseconds.Nanoseconds() {
+		sleepDuration = minWaitTimeInNanoseconds
+		util.Log.Debug("simpleSleepRateLimitStrategy: Reset sleep duration to %f seconds...", sleepDuration.Seconds())
+	}
+	if sleepDuration.Nanoseconds() > maxWaitTimeInNanoseconds.Nanoseconds() {
+		sleepDuration = maxWaitTimeInNanoseconds
+		util.Log.Debug("simpleSleepRateLimitStrategy: Reset sleep duration to %f seconds...", sleepDuration.Seconds())
+	}
+	return sleepDuration
+}
+
+func (s *simpleSleepRateLimitStrategy) convertMicrosecondsToUnixTime(timeInMicroseconds int64) time.Time {
```

Comment:
should be as part of utils?

Reply:
You are right. I'll move this function to `time.go`. :+1:

Reply:
done in aff4dc7